Add blocking checkout with waiter queue

jtdowney opened the pull request:
First, thank you for releasing this. Feel free to ignore this and close it if it doesn't align with your project design. I am working on a system where I need a pool of resources, but I also need to have the requestor block (up to some timeout) if a resource isn't available. I first tried to write it with bath but ran into this issue. I ended up switching to poolboy via some FFI.

Introduce apply_blocking that waits for resources when the pool is exhausted instead of returning NoResourcesAvailable immediately.

jtdowney committed:
Add blocking checkout with waiter queue

Introduce apply_blocking that waits for resources when pool is exhausted
instead of returning NoResourcesAvailable immediately.

jtdowney commented:
Thanks for reviewing, and that is a great point. I can see what you mean about the risk of a leak if the caller remains alive after a checkout timeout. I do still think it would be a valuable path to have a blocking checkout, and the API could change to an explicit panic style like `process.call` with maybe a `_forever` variant. Happy to make an update with that change if you want.

isaacharrisholt commented:
I think that would work great. Would you mind taking a look at implementing that?

jtdowney committed:
Panic on timeout in apply_blocking to fix race condition

Matches process.call behavior - caller dies on timeout, triggering
cleanup via WaiterDown/CallerDown. Removes CheckoutTimedOut error.

jtdowney commented:
I took a crack at the changes in 11221b8. It will now panic and let the caller deal with the situation.

isaacharrisholt commented on src/bath.gleam:
```diff
@@ -502,48 +724,120 @@ fn handle_pool_message(state: State(resource_type), msg: Msg(resource_type)) {
+                    Ok(resource) -> #(
+                      deque.push_back(state.resources, resource),
+                      current_size + 1,
+                    )
```
I think the comments here are either wrong or the logic is

jtdowney replied:
Same as the other comment, this is net zero because 732 does -1 and this does +1. It's a bit confusing, so I want to refactor this to use some smaller helpers.

isaacharrisholt reviewed (changes requested):
Could you also please add a test or two around how the blocking apply works with lazy/eager? The logic seems to have been changed.

jtdowney commented on src/bath.gleam:
```diff
@@ -502,48 +724,120 @@ fn handle_pool_message(state: State(resource_type), msg: Msg(resource_type)) {
+              {
+                // If we create lazily, just decrement the current size - a new resource
+                // will be created when required
+                Lazy -> #(state.resources, current_size)
```
The decrement was moved to 732, so it applied first, and then each branch dealt with the ramifications. I agree that the comment is confusing now. I am working on a cleaner version.

jtdowney committed:
Clarify pool size comments and add blocking strategy tests

jtdowney commented:
I had started down the path of making smaller helpers, and it turned into too much of a refactor for the scope of this PR. So instead, I opted to add the tests and clarify the comments in dcd81c2.

isaacharrisholt commented on src/bath.gleam:
```diff
@@ -502,48 +724,120 @@ fn handle_pool_message(state: State(resource_type), msg: Msg(resource_type)) {
 
           // Shutdown the old resource
           state.shutdown_resource(live_resource.resource)
+          let current_size = state.current_size - 1
```
I'm still not sure I prefer this being pulled out here. I think it makes it more difficult to understand what's happening to the tracked size of the pool in the slightly more complex logic below. What was the reason for doing this one?

isaacharrisholt reviewed (commented):
In theory I'm fine with this PR, but I found myself having to keep a lot in my head when reading it due to the `current_size` changes. If there's a good justification for it, that's fine, but I think the variable should be called `new_size` instead of `current_size`. That said, I'm not sure we need that variable at all.

Apologies for being picky about this, but I'm ultimately going to be the one maintaining this, so I need to make sure I don't get confused 😅

jtdowney committed:
Move current_size calculations back into each branch

jtdowney commented:
No apologies needed, I get it. I agree that `next_size` is a better name for it. I had pulled it out because I found it easier to reason about, but I just sent e2fae45, which factors it back in.

isaacharrisholt committed:
refactor and docs cleanup

isaacharrisholt committed:
refactor waiter rejection

isaacharrisholt reviewed (approved):
I've made a couple of small refactors and made the cleanup logic a little more robust. 

Thank you for this!
